File: ThePlugin/vs/VSJira/api/JiraServerFacade.cs
using System;
using System.Collections.Generic;
using PaZu.api.soap;

namespace PaZu.api
{
    public class JiraServerFacade
    {
        private readonly SortedDictionary<string, SoapSession> sessionMap = new SortedDictionary<string, SoapSession>();

        private static readonly JiraServerFacade INSTANCE = new JiraServerFacade();

        public static JiraServerFacade Instance { get { return INSTANCE; } }

        private JiraServerFacade()
        {
        }

        private SoapSession getSoapSession(JiraServer server)
        {
            SoapSession s;
            if (!sessionMap.TryGetValue(server.Url + server.UserName, out s))
            {
                s = new SoapSession(server.Url);
                s.login(server.UserName, server.Password);
                sessionMap.Add(server.Url + server.UserName, s);
            }
            return s;
        }

        public void login(JiraServer server)
        {
            new SoapSession(server.Url).login(server.UserName, server.Password);
        }

        public List<JiraProject> getProjects(JiraServer server)
        {
            return getSoapSession(server).getProjects();
        }

        public List<JiraNamedEntity> getIssueTypes(JiraServer server)
        {
            return getSoapSession(server).getIssueTypes();
        }

        public List<JiraNamedEntity> getIssueTypes(JiraServer server, JiraProject project)
        {
            return getSoapSession(server).getIssueTypes(project);
        }

        public List<JiraSavedFilter> getSavedFilters(JiraServer server)
        {
            return getSoapSession(server).getSavedFilters();
        }

        public List<JiraIssue> getSavedFilterIssues(JiraServer server, JiraSavedFilter filter, int start, int count)
        {
            RssClient rss = new RssClient(server);
            return rss.getSavedFilterIssues(filter.Id, "issuekey", "ASC", start, count);
        }

        public JiraIssue getIssue(JiraServer server, string key)
        {
            RssClient rss = new RssClient(server);
            return rss.getIssue(key);
        }

        public List<JiraNamedEntity> getPriorities(JiraServer server)
        {
            return getSoapSession(server).getPriorities();
        }

        public List<JiraNamedEntity> getStatuses(JiraServer server)
        {
            return getSoapSession(server).getStatuses();
        }

        public void addComment(JiraIssue issue, string comment)
        {
            getSoapSession(issue.Server).addComment(issue, comment);
        }

        public List<JiraNamedEntity> getActionsForIssue(JiraIssue issue)
        {
            return getSoapSession(issue.Server).getActionsForIssue(issue);
        }

        public List<JiraField> getFieldsForAction(JiraIssue issue, int actionId)
        {
            return getSoapSession(issue.Server).getFieldsForAction(issue, actionId);
        }

        public void runIssueActionWithoutParams(JiraIssue issue, JiraNamedEntity action)
        {
            getSoapSession(issue.Server).runIssueActionWithoutParams(issue, action.Id);
        }

        public List<JiraNamedEntity> getComponents(JiraServer server, JiraProject project)
        {
            return getSoapSession(server).getComponents(project);
        }

        public List<JiraNamedEntity> getVersions(JiraServer server, JiraProject project)
        {
            return getSoapSession(server).getVersions(project);
        }
    }
}

File: atlassian-intellij-connector/src/main/java/com/atlassian/theplugin/commons/jira/JiraServerFacade.java
package com.atlassian.theplugin.commons.jira;

import com.atlassian.connector.commons.jira.JIRAAction;
import com.atlassian.connector.commons.jira.JIRAActionField;
import com.atlassian.connector.commons.jira.JIRAIssue;
import com.atlassian.connector.commons.jira.JiraUserNotFoundException;
import com.atlassian.connector.commons.jira.beans.*;
import com.atlassian.connector.commons.jira.rss.JIRAException;
import com.atlassian.theplugin.commons.ServerType;
import com.atlassian.theplugin.commons.jira.api.JiraIssueAdapter;
import com.atlassian.theplugin.commons.remoteapi.ProductServerFacade;
import com.atlassian.theplugin.commons.remoteapi.RemoteApiException;

import java.util.Calendar;
import java.util.Collection;
import java.util.List;

/**
 * @author pmaruszak
 * @date Sep 28, 2009
 */
public interface JiraServerFacade extends ProductServerFacade {

//    List<JiraIssueAdapter> getIssues(JiraServerData jiraServerData, String queryString,
//                                     String sort, String sortOrder, int start, int size) throws JIRAException;
//
//    List<JiraIssueAdapter> getIssues(JiraServerData jiraServerData, List<JIRAQueryFragment> query,
//                                     String sort, String sortOrder, int start, int size) throws JIRAException;
//
//    List<JiraIssueAdapter> getSavedFilterIssues(JiraServerData jiraServerData, List<JIRAQueryFragment> query,
//                                                String sort, String sortOrder, int start, int size) throws JIRAException;
//
    List<JiraIssueAdapter> getIssues(JiraServerData jiraServerData, JiraFilter filter,
                                     String sort, String sortOrder, int start, int size) throws JIRAException;

    List<JiraIssueAdapter> getSavedFilterIssues(JiraServerData jiraServerData, JIRASavedFilter filter,
                                                String sort, String sortOrder, int start, int size) throws JIRAException;

    List<JIRAProject> getProjects(JiraServerData jiraServerData) throws JIRAException;

    List<JIRAConstant> getStatuses(JiraServerData jiraServerData) throws JIRAException;

    List<JIRAConstant> getIssueTypes(JiraServerData jiraServerData) throws JIRAException;

    List<JIRAConstant> getIssueTypesForProject(JiraServerData jiraServerData, long projectId, String projectKey) throws JIRAException;

    List<JIRAConstant> getSubtaskIssueTypes(JiraServerData jiraServerData) throws JIRAException;

    List<JIRAConstant> getSubtaskIssueTypesForProject(JiraServerData jiraServerData, long projectId, String projectKey) throws JIRAException;

    List<JIRAQueryFragment> getSavedFilters(JiraServerData jiraServerData) throws JIRAException;

    List<JIRAComponentBean> getComponents(JiraServerData jiraServerData, String projectKey) throws JIRAException;

    List<JIRAVersionBean> getVersions(JiraServerData jiraServerData, String projectKey) throws JIRAException;

    List<JIRAPriorityBean> getPriorities(JiraServerData jiraServerData) throws JIRAException;

    List<JIRAResolutionBean> getResolutions(JiraServerData jiraServerData) throws JIRAException;

    List<JIRAAction> getAvailableActions(JiraServerData jiraServerData, JIRAIssue issue) throws JIRAException;

    List<JIRAActionField> getFieldsForAction(JiraServerData jiraServerData, JIRAIssue issue, JIRAAction action)
            throws JIRAException;

    void progressWorkflowAction(JiraServerData jiraServerData, JIRAIssue issue, JIRAAction action) throws JIRAException;

    void progressWorkflowAction(JiraServerData jiraServerData, JIRAIssue issue, JIRAAction action,
                                List<JIRAActionField> fields) throws JIRAException;

    void addComment(JiraServerData jiraServerData, String issueKey, String comment) throws JIRAException;

	void addAttachment(JiraServerData jiraServerData, String issueKey, String name, byte[] content) throws JIRAException;

    JiraIssueAdapter createIssue(JiraServerData jiraServerData, JIRAIssue issue) throws JIRAException;

    JiraIssueAdapter getIssue(JiraServerData jiraServerData, String key) throws JIRAException;

    JiraIssueAdapter getIssueDetails(JiraServerData jiraServerData, JIRAIssue issue) throws JIRAException;

    void logWork(JiraServerData jiraServerData, JIRAIssue issue, String timeSpent, Calendar startDate, String comment,
                 boolean updateEstimate, String newEstimate) throws JIRAException;

    void setAssignee(JiraServerData jiraServerData, JIRAIssue issue, String assignee) throws JIRAException;

	void setReporter(JiraServerData jiraServerData, JIRAIssue issue, String reporter) throws JIRAException;

	void setSummary(JiraServerData jiraServerData, JIRAIssue issue, String summary) throws JIRAException;

	void setDescription(JiraServerData jiraServerData, JIRAIssue issue, String description) throws JIRAException;

	void setType(JiraServerData jiraServerData, JIRAIssue issue, String type) throws JIRAException;

	void setPriority(JiraServerData jiraServerData, JIRAIssue issue, String priority) throws JIRAException;

	void setAffectedVersions(JiraServerData jiraServerData, JIRAIssue issue, String[] versions) throws JIRAException;

	void setFixVersions(JiraServerData jiraServerData, JIRAIssue issue, String[] versions) throws JIRAException;

	void setFields(JiraServerData jiraServerData, JIRAIssue issue, List<JIRAActionField> fields) throws JIRAException;

    JIRAUserBean getUser(JiraServerData jiraServerData, String loginName)
            throws JIRAException, JiraUserNotFoundException;

    List<JIRAComment> getComments(JiraServerData jiraServerData, JIRAIssue issue) throws JIRAException;

    Collection<JIRAAttachment> getIssueAttachements(JiraServerData jiraServerData, JIRAIssue issue)
            throws JIRAException;

    void testServerConnection(JiraServerData jiraServerData) throws RemoteApiException;

    ServerType getServerType();

    List<JIRASecurityLevelBean> getSecurityLevels(JiraServerData jiraServerData, String projectKey)
            throws RemoteApiException, JIRAException;
}
